3rdparty/openvx/include/openvx_hal.hpp

```diff
@@ -0,0 +1,196 @@
+#ifndef OPENCV_OPENVX_HAL_HPP_INCLUDED
+#define OPENCV_OPENVX_HAL_HPP_INCLUDED
+
+#include "opencv2/core/hal/interface.h"
+
+#include "VX/vx.h"
+#include "VX/vxu.h"
+
+#include <string>
+
+//==================================================================================================
+// utility
+// ...
+
+#if 0
+#include <cstdio>
+#define PRINT(...) printf(__VA_ARGS__)
+#else
+#define PRINT(...)
+#endif
+
+#if __cplusplus >= 201103L
+#include <chrono>
+struct Tick
+{
+    typedef std::chrono::time_point<std::chrono::steady_clock> point_t;
+    point_t start;
+    point_t point;
+    Tick()
+    {
+        start = std::chrono::steady_clock::now();
+        point = std::chrono::steady_clock::now();
+    }
+    inline int one()
+    {
+        point_t old = point;
+        point = std::chrono::steady_clock::now();
+        return std::chrono::duration_cast<std::chrono::microseconds>(point - old).count();
+    }
+    inline int total()
+    {
+        return std::chrono::duration_cast<std::chrono::microseconds>(std::chrono::steady_clock::now() - start).count();
+    }
+};
+#endif
+
+//==================================================================================================
+// One more OpenVX C++ binding :-)
+// ...
+
+template <typename T>
+struct VX_Traits
+{
+    enum {
+        Type = 0
+    };
+};
+
+template <>
+struct VX_Traits<uchar>
+{
+    enum {
+        Type = VX_DF_IMAGE_U8
+    };
+};
+
+template <>
+struct VX_Traits<ushort>
+{
+    enum {
+        Type = VX_DF_IMAGE_U16
+    };
+};
+
+template <>
+struct VX_Traits<short>
+{
+    enum {
+        Type = VX_DF_IMAGE_S16
+    };
+};
+
+
+struct vxContext;
+struct vxImage;
+struct vxErr;
+
+
+struct vxErr
+{
+    vx_status status;
+    std::string msg;
+    vxErr(vx_status status_, const std::string & msg_) : status(status_), msg(msg_) {}
+    void check()
+    {
+        if (status != VX_SUCCESS)
+            throw *this;
+    }
+    void print()
+    {
+        PRINT("OpenVX HAL impl error: %d (%s)\n", status, msg.c_str());
+    }
+    static void check(vx_context ctx)
+    {
+        vxErr(vxGetStatus((vx_reference)ctx), "context check").check();
+    }
+    static void check(vx_image img)
+    {
+        vxErr(vxGetStatus((vx_reference)img), "image check").check();
+    }
+    static void check(vx_status s)
+    {
+        vxErr(s, "status check").check();
+    }
+};
+
+
+struct vxContext
+{
+    vx_context ctx;
+    static vxContext * getContext();
```

Comment:
May be it make sense to return pointer directly to vx_context? Looks like it is and should be the only item vxContext is used for.